File: web/src/recommendation/engine.ts
import type {
  Candidate,
  DecisionRecord,
  EliminationStage,
  StageId,
} from "../domain/contracts";
import type {
  DecisionSpec,
  RecommendationInput,
  RelaxationId,
} from "./types";


export class NoSafeCandidate extends Error {
  readonly reasonCounts: Readonly<Record<string, number>>;

  constructor(reasonCounts: Readonly<Record<string, number>>) {
    super("No safe candidate survived the required constraints");
    this.name = "NoSafeCandidate";
    this.reasonCounts = reasonCounts;
  }
}

export interface LocalRecommendationResult {
  readonly decision: DecisionRecord;
  readonly ranked: readonly Candidate[];
  readonly winnerReasonCodes: readonly string[];
}

interface FilterResult {
  readonly survivors: readonly Candidate[];
  readonly reasons: Readonly<Record<string, number>>;
}

function filterCandidates(
  candidates: readonly Candidate[],
  reasonFor: (candidate: Candidate) => string | null,
): FilterResult {
  const survivors: Candidate[] = [];
  const reasons: Record<string, number> = {};
  for (const candidate of candidates) {
    const reason = reasonFor(candidate);
    if (reason === null) survivors.push(candidate);
    else reasons[reason] = (reasons[reason] ?? 0) + 1;
  }
  return { survivors, reasons };
}

function stage(
  id: StageId,
  before: readonly Candidate[],
  after: readonly Candidate[],
  reasons: Readonly<Record<string, number>>,
): EliminationStage {
  const survivorIds = new Set(after.map((candidate) => candidate.dish.id));
  return {
    id,
    input_count: before.length,
    survivor_count: after.length,
    reason_counts: Object.fromEntries(
      Object.entries(reasons).sort(([left], [right]) => left.localeCompare(right)),
    ),
    representative_removed_ids: before
      .filter((candidate) => !survivorIds.has(candidate.dish.id))
      .map((candidate) => candidate.dish.id)
      .sort()
      .slice(0, 3),
  };
}

function localizedPriceMinor(
  candidate: Candidate,
  input: RecommendationInput,
  spec: DecisionSpec,
): number {
  return Math.round(candidate.dish.price_minor * spec.locales[input.locale].usd_multiplier);
}

function overlaps(
  candidateMin: number,
  candidateMax: number,
  targetMin: number,
  targetMax: number,
): boolean {
  return candidateMax >= targetMin && candidateMin <= targetMax;
}

function scoreCandidate(
  candidate: Candidate,
  input: RecommendationInput,
  spec: DecisionSpec,
): number {
  const dish = candidate.dish;
  const desired = new Set(input.constraints.desired_taste_tags);
  const tasteMatches = dish.taste_tags.filter((tag) => desired.has(tag)).length;
  let score = tasteMatches * spec.score.taste_match;
  if (input.context.comfort_preference && dish.taste_tags.includes(input.context.comfort_preference)) {
    score += spec.score.comfort_match;
  }
  score += spec.score.confidence[dish.confidence as keyof typeof spec.score.confidence] ?? 0;
  score -= Math.floor(candidate.place.distance_m / spec.score.distance_divisor);
  score -= Math.floor(dish.price_minor / spec.score.price_divisor);
  score -= (input.repetition.dish_ids[dish.id] ?? 0) * spec.score.recent_dish;
  score -= (input.repetition.base_ingredients[dish.base_ingredient] ?? 0)
    * spec.score.recent_base_ingredient;
  const preference = [
    ...dish.cuisine_tags.map((tag) => input.preferences.cuisine[tag] ?? 0),
    ...dish.taste_tags.map((tag) => input.preferences.taste[tag] ?? 0),
  ].reduce((total, value) => total + value, 0);
  score += Math.round(preference * spec.score.preference_scale);
  return score;
}

function stableValue(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(stableValue);
  if (typeof value === "object" && value !== null) {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .sort(([left], [right]) => left.localeCompare(right))
        .map(([key, item]) => [key, stableValue(item)]),
    );
  }
  return value;
}

async function sha256(value: unknown): Promise<string> {
  const bytes = new TextEncoder().encode(JSON.stringify(stableValue(value)));
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return [...new Uint8Array(digest)].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

function canRelax(input: RecommendationInput, relaxation: RelaxationId): boolean {
  return input.constraints.allowed_relaxations.includes(relaxation);
}

function softStage(
  id: Extract<StageId, "repetition" | "taste" | "duration" | "budget">,
  current: readonly Candidate[],
  filtered: FilterResult,
  relaxation: RelaxationId,
  input: RecommendationInput,
  relaxations: RelaxationId[],
): { readonly survivors: readonly Candidate[]; readonly trace: EliminationStage } {
  if (filtered.survivors.length > 0) {
    return {
      survivors: filtered.survivors,
      trace: stage(id, current, filtered.survivors, filtered.reasons),
    };
  }
  if (!canRelax(input, relaxation)) throw new NoSafeCandidate(filtered.reasons);
  relaxations.push(relaxation);
  return {
    survivors: current,
    trace: stage(id, current, current, filtered.reasons),
  };
}

function winnerReasons(winner: Candidate, input: RecommendationInput): readonly string[] {
  const reasons: string[] = [];
  if (winner.dish.taste_tags.some((tag) => input.constraints.desired_taste_tags.includes(tag))) {
    reasons.push("taste_match");
  }
  if (
    input.constraints.minimum_protein_g !== null
    && winner.dish.protein_g.min >= input.constraints.minimum_protein_g
  ) reasons.push("protein_match");
  if ((input.repetition.dish_ids[winner.dish.id] ?? 0) === 0) reasons.push("recent_variety");
  if (winner.dish.estimated_minutes <= input.constraints.max_duration_minutes) {
    reasons.push("duration_match");
  }
  reasons.push("confidence_match");
  return reasons.slice(0, 3);
}

export async function recommendLocal(
  candidates: readonly Candidate[],
  input: RecommendationInput,
  spec: DecisionSpec,
  now: Date,
): Promise<LocalRecommendationResult> {
  const ordered = [...candidates].sort((left, right) => left.dish.id.localeCompare(right.dish.id));
  if (ordered.length === 0) throw new NoSafeCandidate({ empty_catalog: 1 });
  const stages: EliminationStage[] = [stage("found", ordered, ordered, {})];
  const relaxations: RelaxationId[] = [];

  const available = filterCandidates(ordered, (candidate) => {
    if (candidate.place.open_state === "closed") return "place_closed";
    if (!candidate.place.order_destination) return "missing_order_destination";
    return null;
  });
  stages.push(stage("available", ordered, available.survivors, available.reasons));
  if (available.survivors.length === 0) throw new NoSafeCandidate(available.reasons);

  const allergens = new Set(input.constraints.excluded_allergens);
  const ingredients = new Set(
    input.constraints.excluded_ingredients.map((value) => value.toLocaleLowerCase()),
  );
  const sessionExclusions = new Set(input.session_exclusions);
  const safe = filterCandidates(available.survivors, (candidate) => {
    if (sessionExclusions.has(candidate.dish.id)) return "session_excluded";
    if ([...candidate.dish.allergens, ...candidate.dish.possible_allergens]
      .some((allergen) => allergens.has(allergen))) return "allergen_excluded";
    if (candidate.dish.ingredients.some((ingredient) => ingredients.has(ingredient.toLocaleLowerCase()))) {
      return "ingredient_excluded";
    }
    return null;
  });
  stages.push(stage("safety", available.survivors, safe.survivors, safe.reasons));
  if (safe.survivors.length === 0) throw new NoSafeCandidate(safe.reasons);

  let nutrition = safe.survivors;
  const nutritionReasons: Record<string, number> = {};
  const targetEnergy = input.context.energy_range_kcal;
  if (targetEnergy) {
    const energy = filterCandidates(nutrition, (candidate) => overlaps(
      candidate.dish.energy_kcal.min,
      candidate.dish.energy_kcal.max,
      targetEnergy.min,
      targetEnergy.max,
    ) ? null : "energy_outside_range");
    Object.assign(nutritionReasons, energy.reasons);
    if (energy.survivors.length > 0) nutrition = energy.survivors;
  }
  if (input.constraints.minimum_protein_g !== null) {
    const proteinFloor = input.constraints.minimum_protein_g;
    const protein = filterCandidates(nutrition, (candidate) => (
      candidate.dish.protein_g.min >= proteinFloor ? null : "protein_below_floor"
    ));
    Object.assign(nutritionReasons, protein.reasons);
    if (protein.survivors.length > 0) nutrition = protein.survivors;
  }
  stages.push(stage("nutrition", safe.survivors, nutrition, nutritionReasons));

  const repeated = filterCandidates(nutrition, (candidate) => {
    if (input.context.recent_categories_to_avoid.includes(candidate.dish.base_ingredient)) {
      return "recent_repetition";
    }
    if ((input.repetition.dish_ids[candidate.dish.id] ?? 0) >= 1) return "recent_repetition";
    if ((input.repetition.base_ingredients[candidate.dish.base_ingredient] ?? 0) >= 2) {
      return "recent_repetition";
    }
    return null;
  });
  let soft = softStage(
    "repetition",
    nutrition,
    repeated,
    "recent_repetition",
    input,
    relaxations,
  );
  stages.push(soft.trace);

  const desired = new Set(input.constraints.desired_taste_tags);
  const taste = filterCandidates(soft.survivors, (candidate) => (
    desired.size === 0 || candidate.dish.taste_tags.some((tag) => desired.has(tag))
      ? null
      : "taste_mismatch"
  ));
  soft = softStage("taste", soft.survivors, taste, "taste", input, relaxations);
  stages.push(soft.trace);

  const duration = filterCandidates(soft.survivors, (candidate) => (
    candidate.dish.estimated_minutes <= input.constraints.max_duration_minutes
      ? null
      : "too_slow"
  ));
  soft = softStage("duration", soft.survivors, duration, "duration", input, relaxations);
  stages.push(soft.trace);

  const budget = filterCandidates(soft.survivors, (candidate) => (
    localizedPriceMinor(candidate, input, spec) <= input.constraints.max_price_minor
      ? null
      : "over_budget"
  ));
  soft = softStage("budget", soft.survivors, budget, "budget", input, relaxations);
  stages.push(soft.trace);

  const ranked = [...soft.survivors].sort((left, right) => (
    scoreCandidate(right, input, spec) - scoreCandidate(left, input, spec)
    || left.place.distance_m - right.place.distance_m
    || left.dish.price_minor - right.dish.price_minor
    || left.dish.id.localeCompare(right.dish.id)
  ));
  const winner = ranked[0];
  if (!winner) throw new NoSafeCandidate({ empty_ranked_set: 1 });
  const reserve = ranked[1] ?? null;
  stages.push(stage(
    "winner",
    ranked,
    [winner],
    ranked.length > 1 ? { reserve_or_lower_score: ranked.length - 1 } : {},
  ));

  const inputHash = await sha256({
    candidates: ordered,
    input,
    rules_version: spec.version,
  });
  return {
    decision: {
      decision_id: `decision-${inputHash.slice(0, 16)}`,
      engine_version: "engine.v2",
      catalog_version: "catalog.v1",
      input_sha256: inputHash,
      created_at: now.toISOString(),
      stages,
      winner_id: winner.dish.id,
      reserve_id: reserve?.dish.id ?? null,
      relaxations,
    },
    ranked,
    winnerReasonCodes: winnerReasons(winner, input),
  };
}

File: backend/src/onedish_api/engine.py
"""Canonical deterministic meal-elimination engine."""

from __future__ import annotations

import hashlib
import json
from collections import Counter
from dataclasses import asdict
from datetime import datetime
from math import ceil
from typing import Callable, Sequence

from onedish_api.domain import (
    Candidate,
    Confidence,
    DecisionRecord,
    EliminationStage,
    MealConstraints,
    MealContext,
)
from onedish_api.history import PreferenceWeights, RepetitionProfile


class NoSafeCandidate(ValueError):
    def __init__(self, reason_counts: dict[str, int]) -> None:
        super().__init__("no candidate survived the required constraints")
        self.reason_counts = reason_counts


def _stage(
    stage_id: str,
    before: Sequence[Candidate],
    after: Sequence[Candidate],
    reasons: dict[str, int],
) -> EliminationStage:
    after_ids = {candidate.dish.id for candidate in after}
    removed = tuple(
        sorted(candidate.dish.id for candidate in before if candidate.dish.id not in after_ids)[:3]
    )
    return EliminationStage(
        id=stage_id,  # type: ignore[arg-type]
        input_count=len(before),
        survivor_count=len(after),
        reason_counts=dict(sorted(reasons.items())),
        representative_removed_ids=removed,
    )


def _filter(
    candidates: Sequence[Candidate],
    reason_for: Callable[[Candidate], str | None],
) -> tuple[list[Candidate], dict[str, int]]:
    survivors: list[Candidate] = []
    reasons: Counter[str] = Counter()
    for candidate in candidates:
        reason = reason_for(candidate)
        if reason is None:
            survivors.append(candidate)
        else:
            reasons[reason] += 1
    return survivors, dict(reasons)


def _overlaps(candidate_min: int, candidate_max: int, target_min: int, target_max: int) -> bool:
    return candidate_max >= target_min and candidate_min <= target_max


def _confidence_rank(confidence: Confidence) -> int:
    return {
        Confidence.authoritative: 4,
        Confidence.high: 3,
        Confidence.medium: 2,
        Confidence.low: 1,
    }[confidence]


def _score(
    candidate: Candidate,
    context: MealContext,
    constraints: MealConstraints,
    repetition: RepetitionProfile,
    preferences: PreferenceWeights,
) -> int:
    dish = candidate.dish
    desired = set(constraints.desired_taste_tags)
    taste_match = len(desired.intersection(dish.taste_tags))
    score = taste_match * 1_000
    if context.comfort_preference and context.comfort_preference in dish.taste_tags:
        score += 450
    score += _confidence_rank(dish.confidence) * 220
    score -= candidate.place.distance_m // 5
    score -= dish.price_minor // 15
    score -= repetition.dish_ids.get(dish.id, 0) * 800
    score -= repetition.base_ingredients.get(dish.base_ingredient, 0) * 350
    score += round(
        1_000
        * (
            sum(preferences.cuisine.get(tag, 0) for tag in dish.cuisine_tags)
            + sum(preferences.taste.get(tag, 0) for tag in dish.taste_tags)
        )
    )
    return score


def _canonical_hash(
    candidates: Sequence[Candidate],
    context: MealContext,
    constraints: MealConstraints,
    repetition: RepetitionProfile,
    preferences: PreferenceWeights,
    catalog_version: str,
) -> str:
    payload = {
        "candidates": [
            candidate.model_dump(mode="json")
            for candidate in sorted(candidates, key=lambda item: item.dish.id)
        ],
        "context": context.model_dump(mode="json"),
        "constraints": constraints.model_dump(mode="json"),
        "repetition": asdict(repetition),
        "preferences": asdict(preferences),
        "catalog_version": catalog_version,
        "engine_version": "engine.v1",
    }
    raw = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    return hashlib.sha256(raw).hexdigest()


def recommend(
    candidates: Sequence[Candidate],
    context: MealContext,
    constraints: MealConstraints,
    repetition: RepetitionProfile,
    preferences: PreferenceWeights,
    *,
    catalog_version: str,
    created_at: datetime,
) -> DecisionRecord:
    ordered = sorted(candidates, key=lambda candidate: candidate.dish.id)
    if not ordered:
        raise NoSafeCandidate({"empty_catalog": 1})
    stages: list[EliminationStage] = [
        _stage("found", ordered, ordered, {})
    ]
    relaxations: list[str] = []

    available, reasons = _filter(
        ordered,
        lambda item: (
            "place_closed"
            if item.place.open_state == "closed"
            else "missing_order_destination"
            if not item.place.order_destination
            else None
        ),
    )
    stages.append(_stage("available", ordered, available, reasons))
    if not available:
        raise NoSafeCandidate(reasons)

    excluded_allergens = set(constraints.excluded_allergens)
    excluded_ingredients = {value.casefold() for value in constraints.excluded_ingredients}

    def safety_reason(item: Candidate) -> str | None:
        dish = item.dish
        if excluded_allergens.intersection((*dish.allergens, *dish.possible_allergens)):
            return "allergen_excluded"
        ingredients = {ingredient.casefold() for ingredient in dish.ingredients}
        if excluded_ingredients.intersection(ingredients):
            return "ingredient_excluded"
        if constraints.max_price_minor is not None and dish.price_minor > constraints.max_price_minor:
            return "over_budget"
        if dish.currency != constraints.currency:
            return "currency_mismatch"
        return None

    safe, reasons = _filter(available, safety_reason)
    stages.append(_stage("safety_budget", available, safe, reasons))
    if not safe:
        raise NoSafeCandidate(reasons)

    nutrition_input = safe
    nutrition = safe
    nutrition_reasons: Counter[str] = Counter()
    target_energy = constraints.energy_range_kcal or context.energy_range_kcal
    if target_energy is not None:
        energy, energy_reasons = _filter(
            nutrition,
            lambda item: (
                None
                if _overlaps(
                    item.dish.energy_kcal.min,
                    item.dish.energy_kcal.max,
                    target_energy.min,
                    target_energy.max,
                )
                else "energy_outside_range"
            ),
        )
        nutrition_reasons.update(energy_reasons)
        if energy:
            nutrition = energy
        elif "energy_range" in constraints.allowed_relaxations:
            relaxations.append("energy_range")
        else:
            stages.append(_stage("nutrition", nutrition_input, [], dict(nutrition_reasons)))
            raise NoSafeCandidate(dict(nutrition_reasons))

    protein_floor = constraints.minimum_protein_g
    if protein_floor is not None:
        protein, protein_reasons = _filter(
            nutrition,
            lambda item: (
                None if item.dish.protein_g.min >= protein_floor else "protein_below_floor"
            ),
        )
        nutrition_reasons.update(protein_reasons)
        if protein:
            nutrition = protein
        elif "protein_floor" in constraints.allowed_relaxations:
            relaxations.append("protein_floor")
        else:
            stages.append(_stage("nutrition", nutrition_input, [], dict(nutrition_reasons)))
            raise NoSafeCandidate(dict(nutrition_reasons))
    stages.append(_stage("nutrition", nutrition_input, nutrition, dict(nutrition_reasons)))

    avoid = set(context.recent_categories_to_avoid)
    repeated = {
        item.dish.id
        for item in nutrition
        if item.dish.base_ingredient in avoid
        or repetition.base_ingredients.get(item.dish.base_ingredient, 0) >= 2
        or repetition.dish_ids.get(item.dish.id, 0) >= 1
    }
    non_repeated = [item for item in nutrition if item.dish.id not in repeated]
    if non_repeated:
        repetition_survivors = non_repeated
        repetition_reasons = {"recent_repetition": len(repeated)} if repeated else {}
    else:
        repetition_survivors = nutrition
        repetition_reasons = {"recent_repetition_retained": len(repeated)} if repeated else {}
        if repeated and "recent_repetition" in constraints.allowed_relaxations:
            relaxations.append("recent_repetition")
    stages.append(
        _stage("repetition", nutrition, repetition_survivors, repetition_reasons)
    )

    ranked = sorted(
        repetition_survivors,
        key=lambda item: (
            -_score(item, context, constraints, repetition, preferences),
            -_confidence_rank(item.dish.confidence),
            item.place.distance_m,
            item.dish.price_minor,
            item.dish.id,
        ),
    )
    shortlist_size = min(len(ranked), max(2, ceil(len(ranked) * 0.35)))
    shortlist = ranked[:shortlist_size]
    stages.append(
        _stage(
            "taste_confidence",
            repetition_survivors,
            shortlist,
            {"lower_score": len(ranked) - len(shortlist)} if len(ranked) > len(shortlist) else {},
        )
    )
    winner = shortlist[0]
    reserve = shortlist[1] if len(shortlist) > 1 else None
    stages.append(
        _stage(
            "winner",
            shortlist,
            [winner],
            {"reserve_or_lower_score": len(shortlist) - 1} if len(shortlist) > 1 else {},
        )
    )

    input_hash = _canonical_hash(
        ordered, context, constraints, repetition, preferences, catalog_version
    )
    return DecisionRecord(
        decision_id=f"decision-{input_hash[:16]}",
        engine_version="engine.v1",
        catalog_version=catalog_version,
        input_sha256=input_hash,
        created_at=created_at,
        stages=tuple(stages),
        winner_id=winner.dish.id,
        reserve_id=reserve.dish.id if reserve else None,
        relaxations=tuple(relaxations),
    )
